app/uk/gov/hmrc/eusubsidycompliancefrontend/controllers/UndertakingController.scala

```diff
@@ -146,14 +153,107 @@ class UndertakingController @Inject() (
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  def postConfirmEmail: Action[AnyContent] = withCDSAuthenticatedUser.async { implicit request =>
+    implicit val eori: EORI = request.eoriNumber
+    val verifiedEmail = for {
+      stored <- emailVerificationService.getEmailVerification(request.eoriNumber)
+      cds <- emailService.retrieveEmailByEORI(request.eoriNumber)
+      result = if(stored.isDefined) stored.get.email.some else cds match {
+        case RetrieveEmailResponse(VerifiedEmail, Some(value)) => value.value.some
+        case _ => Option.empty
+      }
+    } yield result
+
+    verifiedEmail flatMap {
+      case Some(email) =>
+        optionalEmailForm
+          .bindFromRequest()
+          .fold(
+            errors => BadRequest(confirmEmailPage(errors, EmailAddress(email), routes.EligibilityController.getCustomsWaivers().url)).toFuture,
+            {
+              case OptionalStringFormInput("true", None) => {
+                for {
+                  _ <- emailVerificationService.verifyEori(request.eoriNumber)
+                  _ <- store.update[UndertakingJourney](_.setVerifiedEmail(email))
+                } yield (Redirect(routes.UndertakingController.getCheckAnswers()))
+              }
+              case OptionalStringFormInput("false", Some(email)) => {
+                val pendingVerificationId = UUID.randomUUID().toString
+                for {
+                  _ <- emailVerificationService.addVerificationRequest(request.eoriNumber, email, pendingVerificationId)
+                  verificationResponse <- emailVerificationService.verifyEmail(request.authorityId, email, pendingVerificationId)
+                } yield emailVerificationService.emailVerificationRedirect(verificationResponse)
+              }
+              case _ => Redirect(routes.EligibilityController.getNotEligible()).toFuture
+            }
+          )
+      case _ => emailForm.bindFromRequest().fold(
+        errors => BadRequest(inputEmailPage(errors, routes.EligibilityController.getCustomsWaivers().url)).toFuture,
+        form => {
+          val pendingVerificationId = UUID.randomUUID().toString
+          for {
+            _ <- emailVerificationService.addVerificationRequest(request.eoriNumber, form.value, pendingVerificationId)
+            verificationResponse <- emailVerificationService.verifyEmail(request.authorityId, form.value, pendingVerificationId)
+          } yield emailVerificationService.emailVerificationRedirect(verificationResponse)
+        }
+      )
+    }
+  }
+
+  def getVerifyEmail(pendingVerificationId: String): Action[AnyContent] = withCDSAuthenticatedUser.async { implicit request =>
+    implicit val eori: EORI = request.eoriNumber
+    store.get[UndertakingJourney].flatMap {
+      case Some(journey) =>
+        for {
+          e <- emailVerificationService.verifyVerificationRequest(request.eoriNumber, pendingVerificationId)
+          wasSuccessful = e.getMatchedCount > 0
```

Comment:
Are we going to have mulitple occurrences of the same email address? I figured we could use `eori, email` as a unique key but I could be wrong.

Reply:
We decided that it's just one verified email per EORI. Good catch though!